src/NHibernate/Engine/Query/HQLQueryPlan.cs

```diff
@@ -174,6 +198,69 @@ public IEnumerable PerformIterate(QueryParameters queryParameters, IEventSource
 			return new JoinedEnumerable(results);
 		}
 
+		// Since v5.3
+		[Obsolete("This method has no more usages and will be removed in a future version")]
+		public async Task<IEnumerable> PerformIterateAsync(QueryParameters queryParameters, IEventSource session, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			if (Log.IsDebugEnabled())
+			{
+				Log.Debug("enumerable: {0}", _sourceQuery);
+				queryParameters.LogParameters(session.Factory);
+			}
+			if (Translators.Length == 0)
+			{
+				return CollectionHelper.EmptyEnumerable;
+			}
+			if (Translators.Length == 1)
+			{
+				return await (Translators[0].GetEnumerableAsync(queryParameters, session, cancellationToken)).ConfigureAwait(false);
+			}
+			var results = new IEnumerable[Translators.Length];
+			for (int i = 0; i < Translators.Length; i++)
+			{
+				var result = await (Translators[i].GetEnumerableAsync(queryParameters, session, cancellationToken)).ConfigureAwait(false);
+				results[i] = result;
+			}
+			return new JoinedEnumerable(results);
+		}
+
+		// Since v5.3
+		[Obsolete("This method has no more usages and will be removed in a future version")]
+		public async Task<IEnumerable<T>> PerformIterateAsync<T>(QueryParameters queryParameters, IEventSource session, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			return new SafetyEnumerable<T>(await (PerformIterateAsync(queryParameters, session, cancellationToken)).ConfigureAwait(false));
+		}
+
+		public IAsyncEnumerable<T> PerformAsyncIterate<T>(QueryParameters queryParameters, IEventSource session)
+		{
+			if (Log.IsDebugEnabled())
+			{
+				Log.Debug("enumerable: {0}", _sourceQuery);
+				queryParameters.LogParameters(session.Factory);
+			}
+
+			if (Translators.Length == 0)
+			{
+				return new CollectionHelper.EmptyAsyncEnumerable<T>();
```

Comment:
I don't think we need this optimization along with `CollectionHelper.EmptyAsyncEnumerable` (at least not in this PR). This case means query is executed  on not mapped entity. If it's not throwing already before - I see no point in optimizing broken case.

Reply:
`CollectionHelper.EmptyAsyncEnumerable` is now removed thanks to your proposal for yielding `IAsyncEnumerable`.